src/server/game/Entities/Unit/StatSystem.cpp

```diff
@@ -1068,19 +1016,61 @@ void Guardian::UpdateResistances(uint32 school)
 
 void Guardian::UpdateArmor()
 {
-    float value = 0.0f;
-    float bonus_armor = 0.0f;
     UnitMods unitMod = UNIT_MOD_ARMOR;
 
-    // hunter pets gain 35% of owner's armor value, warlock pets gain 100% of owner's armor
-    if (IsHunterPet())
-        bonus_armor = float(CalculatePct(m_owner->GetArmor(), 70));
-    else if (IsPet())
-        bonus_armor = m_owner->GetArmor();
+    float armor = GetArmor();
+    float value = 0.0f;
+    float pctFromOwnerArmor = 0.0f;
 
-    value  = GetModifierValue(unitMod, BASE_VALUE);
+    PetType petType = IsHunterPet() ? HUNTER_PET : SUMMON_PET;
+
+    switch (petType)
+    {
+        case SUMMON_PET:
+            switch (GetEntry())
+            {
+                // Mage
+                case ENTRY_WATER_ELEMENTAL:
+                    pctFromOwnerArmor = 300.f;
+                    break;
+                // Warlock
+                case ENTRY_FELGUARD:
+                case ENTRY_VOIDWALKER:
+                    pctFromOwnerArmor = 400.f;
+                    break;
+                case ENTRY_FELHUNTER:
+                case ENTRY_IMP:
+                case ENTRY_SUCCUBUS:
+                    pctFromOwnerArmor = 300.f;
+                    break;
+                case ENTRY_NIUZAO:
+                    pctFromOwnerArmor = 400.f;
```

Comment:
Besides the obvious uglyness related to the abundance of magic valueness: missing break.